src/relion/__init__.py

```diff
@@ -178,34 +197,125 @@ def __init__(self):
         self._seen_before = []
         self._fresh_called = False
         self._cache = {}
+        self._validation_cache = {}
+
+    @staticmethod
+    def _update_temp_validation_cache(
+        stage, job, current_result, cache, results_for_validation
+    ):
+        validation_check = stage.for_validation(current_result)
+        if validation_check:
+            if not cache.get(stage):
+                cache[stage] = {}
+            if (stage, job) not in results_for_validation:
+                results_for_validation.append((stage, job))
+            cache[stage].update(validation_check)
 
     def consume(self, results):
         self._results = [(r[0], r[1]) for r in results]
+        curr_val_cache = {}
+        results_for_validation = []
         if not self._fresh_called:
             self._fresh_results = self._results
             for r in results:
-                if r[2] not in [p[1] for p in self._seen_before]:
+                end_time_not_seen_before = r[2] not in [p[1] for p in self._seen_before]
+                if end_time_not_seen_before:
                     self._cache[r[1]] = []
-                    for single_res in r[0][r[1]]:
+                for single_res in r[0][r[1]]:
+                    self._update_temp_validation_cache(
+                        r[0], r[1], single_res, curr_val_cache, results_for_validation
+                    )
+                    if end_time_not_seen_before:
                         self._cache[r[1]].append(r[0].for_cache(single_res))
                 if (r[1], r[2]) not in self._seen_before:
                     self._seen_before.append((r[1], r[2]))
```

Comment:
this seems very abstract and is hard to parse. Maybe `r`s should be `namedtuple`s?

Reply:
I agree that this stuff is a mess and needs a re-work. I'll have a think about whether there is a way that I can make it neater quickly but if it isn't easy I'll add an issue or something to remember to improve it in the future

Reply:
I replaced the tuple `r`s with `namedtuple`s as you suggested and I think it's a big improvement, thanks. I used two different `namedtuple`s as in some cases I want to store the Relion stage object along with the job directory and last completion time for that job, while in other cases I drop the stage object. I think this stuff will probably get some sort of overhaul at some point.